src/main/ruby/core/process_mirror.rb

```diff
@@ -387,7 +416,11 @@ def redirect_file_descriptor(from, to)
         end
 
         def spawn(options, command, arguments)
-          pid = Truffle::Process.spawn command, arguments, env_array, options
+          pid = begin
+                  Truffle::Process.spawn command, arguments, env_array, options
+                ensure
+                  @files_for_child.each { |f| f.close }
+                end
```

Comment:
Unfortunately this will not close the fds if an error happens during
```ruby
exe = Execute.new(*args)
exe.spawn_setup(false)
```
above.
So the ensure should be moved up there, and probably we can add a method `close` or so to `Execute` to do the cleanup.

Reply:
Slightly more subtle than that sadly, because the error can occur in `Execute#initialize` in which case you don't get an object back to call `close_files` on..

Reply:
Yeah, we should really not have `#initialize` do the option parsing.